### ISSUE TITLE
Allow access to the object instances created via the data attributes API inside JS (usage example problems with the modal component)

### ISSUE THREAD

**Issue by KinG-InFeT** [themesberg/flowbite#119]
**Describe the bug**
if open modal with javascript, i can't close modal with javascript helper or data-modal-toggle=""

**To Reproduce**
1. Go to official site: https://flowbite.com/docs/components/modal/#methods
2. open modal with javascript  from chrome console
3. ``` new Modal(document.getElementById('defaultModal'), null).show(); ```
4. Close modal with (X) icon inside modal, not close modal!
5. no javscript error

**Expected behavior**
close correctly modal

**Screenshots**
![image](https://user-images.githubusercontent.com/193969/159908723-e48d4648-b945-4780-b0f3-5db3d38d8cb9.png)

**Desktop (please complete the following information):**
 - OS: Ubuntu
 - Browser chrome
 - Version 1.4.1

**Smartphone (please complete the following information):**

**Additional context**

**Comment by zoltanszogyenyi** [comment]
Hey @KinG-InFeT,

If you use the object methods you need to add the event listeners yourself on the close icon to close the same instance of the Modal.

For example:

```
const modal = new Modal(document.getElementById('defaultModal'), null).show();
const icon = document.getElementById('close-icon');

icon.addEventListener('click', () => {
    modal.hide();
})
```

**Comment by KinG-InFeT** [comment]
hi @zoltanszogyenyi ,
thank you for reply, after hide, the mouseenter or hover event continue to work. can i disable tooltip permanently before reuse .show()?

**Comment by KinG-InFeT** [comment]
the background is: modal-backdrop not removed by javascript

**Comment by KinG-InFeT** [comment]
@zoltanszogyenyi i resolve my problem with small javascript code (very very simple) with jquery
```
$('[data-modal-toggle]').on('click', function (event) {
        var modal_id = $(this).data("modal-toggle");
        if($('#'+modal_id).hasClass("hidden")) {
            $('[modal-backdrop]').remove();
        }
    });
```

**Comment by michelvermeulen** [comment]
Hi guys.

Coming from bootstrap modals (pure HTML and Vue), this seems hacky. If I initialize the modal via JS, and toggle it via JS, and then click on a button with data-modal-toggle, the background stays on.

```
modal = new Modal(targetEl, options);
    modal.show();
```

```
<button type="button"
          class="text-gray-400 bg-transparent hover:bg-gray-200 hover:text-gray-900 rounded-lg text-sm p-1.5 ml-auto inline-flex items-center dark:hover:bg-gray-600 dark:hover:text-white"
          data-modal-toggle="{{$id}}">
          <svg class="w-5 h-5" fill="currentColor" viewBox="0 0 20 20" xmlns="http://www.w3.org/2000/svg">
            <path fill-rule="evenodd"
              d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
              clip-rule="evenodd"></path>
          </svg>
        </button>
```

![CleanShot 2022-04-03 at 08 26 39](https://user-images.githubusercontent.com/1824611/161414852-0cac31e2-2ced-44ca-8825-f2cab3676c67.gif)

I ended up using custom data-modal-hide and data-modal-show to make this work. Too bad though the toggle doesn't work :/

**Comment by michelvermeulen** [comment]
Alright so looking at the sourcecode, it seems like the lib is automatically creating an instance of the modal via the data-modal-toggle attribute, and then manually creating the modal via custom JS creates another instance of the same modal.

Are there any instructions on how to toggle a modal in the JS without re-creating a new modal instance?

**Comment by mpapado3** [comment]
I am having also the same issue, the modal is created with js and when it is closed through any button the backdrop stays on.
Didn't understood @michelvermeulen solution with data-modal-hide and data-modal-show.

**Comment by csoutham** [comment]
I've just pushed a pull request to add an attribute so modals can be opened by default, without the need for any JS. Fingers crossed it accepted.
https://github.com/themesberg/flowbite/pull/141

**Comment by zoltanszogyenyi** [comment]
@csoutham merged the PR. I think it's safe to close this issue :)

**Comment by KinG-InFeT** [comment]
> @csoutham merged the PR. I think it's safe to close this issue :)

hello, the PR however does not solve the problem in the ticket, it only helps not to open the modal via javascript, avoiding the basic problem. I believe the bug will still have to be fixed.

**Comment by csoutham** [comment]
@KinG-InFeT If you use the PR that I pushed, with the data-attribute for showing the modal be default, it'll add the background programmatically and remove it when toggled.

**Comment by KinG-InFeT** [comment]
any update?

**Comment by maninthemirror** [comment]
Same problem

**Comment by lfsaga** [comment]
Same for me

**Comment by Rodrigo-lpds** [comment]
Same here

**Comment by harrisonratcliffe** [comment]
Same problem here

**Comment by zoltanszogyenyi** [comment]
Hey everyone,

When opening the modal with JavaScript, please use the `modal.hide();` method by manually adding the event listener to the close button.

The `data-modal-toggle` will only work when the modal is initialized only via the data attributes.

Here's an example:

```
// set the modal menu element
const targetEl = document.getElementById('modalEl');

// options with default values
const options = {
  placement: 'bottom-right',
  backdropClasses: 'bg-gray-900 bg-opacity-50 dark:bg-opacity-80 fixed inset-0 z-40',
  onHide: () => {
      console.log('modal is hidden');
  },
  onShow: () => {
      console.log('modal is shown');
  },
  onToggle: () => {
      console.log('modal has been toggled');
  }
};
```

Create a new Modal object:

```
/*
* targetEl: required
* options: optional
*/
const modal = new Modal(targetEl, options);
```

Use these methods to show or hide the modal:

```
// show the modal
modal.show();

// hide the modal
modal.hide();
```

So in this case you can do something like:

```
<button id="close-modal">x</button>
```

Then do this:

```
document.getElementById('close-modal').addEventListener('click', function () {
    modal.hide();
})

Read more about the `[Modal JS usage on Flowbite](https://flowbite.com/docs/components/modal/#javascript-behaviour)`.

**Comment by harrisonratcliffe** [comment]
> > Hey everyone,
> > When opening the modal with JavaScript, please use the `modal.hide();` method by manually adding the event listener to the close button.
> > The `data-modal-toggle` will only work when the modal is initialized only via the data attributes.
> > Here's an example:
> > ```
> > // set the modal menu element
> > const targetEl = document.getElementById('modalEl');
> > 
> > // options with default values
> > const options = {
> >   placement: 'bottom-right',
> >   backdropClasses: 'bg-gray-900 bg-opacity-50 dark:bg-opacity-80 fixed inset-0 z-40',
> >   onHide: () => {
> >       console.log('modal is hidden');
> >   },
> >   onShow: () => {
> >       console.log('modal is shown');
> >   },
> >   onToggle: () => {
> >       console.log('modal has been toggled');
> >   }
> > };
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Create a new Modal object:
> > ```
> > /*
> > * targetEl: required
> > * options: optional
> > */
> > const modal = new Modal(targetEl, options);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Use these methods to show or hide the modal:
> > ```
> > // show the modal
> > modal.show();
> > 
> > // hide the modal
> > modal.hide();
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > So in this case you can do something like:
> > ```
> > <button id="close-modal">x</button>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Then do this:
> > ```
> > document.getElementById('close-modal').addEventListener('click', function () {
> >     modal.hide();
> > })
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Read more about the [Modal JS usage on Flowbite](https://flowbite.com/docs/components/modal/#javascript-behaviour).
> 
> hello, unfortunately we still do not understand the basic problem. use javascript to close the modal we agree, but if you open it with a normal toggle from a button and then you want to manage the closure with javascript, the reference to "modal" is lost (not to mention the possibility of having more modals open at the same time, how do you manage the reference? the problem still has not been solved and the management of events and the library should be improved to give the possibility to manage the modal in both ways (the correct motion in the management of the modal is exactly as bootstrap does https://getbootstrap.com/docs/5.2/components/modal/). in a nutshell, the backdrop must have the reference to the modal in some way, to be able to fully manage it!

This exactly.

**Comment by harrisonratcliffe** [comment]
> Hi guys.
> 
> Coming from bootstrap modals (pure HTML and Vue), this seems hacky. If I initialize the modal via JS, and toggle it via JS, and then click on a button with data-modal-toggle, the background stays on.
> 
> ```
> modal = new Modal(targetEl, options);
>     modal.show();
> ```
> 
> ```
> <button type="button"
>           class="text-gray-400 bg-transparent hover:bg-gray-200 hover:text-gray-900 rounded-lg text-sm p-1.5 ml-auto inline-flex items-center dark:hover:bg-gray-600 dark:hover:text-white"
>           data-modal-toggle="{{$id}}">
>           <svg class="w-5 h-5" fill="currentColor" viewBox="0 0 20 20" xmlns="http://www.w3.org/2000/svg">
>             <path fill-rule="evenodd"
>               d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
>               clip-rule="evenodd"></path>
>           </svg>
>         </button>
> ```
> 
> ![CleanShot 2022-04-03 at 08 26 39](https://user-images.githubusercontent.com/1824611/161414852-0cac31e2-2ced-44ca-8825-f2cab3676c67.gif) [ ](https://user-images.githubusercontent.com/1824611/161414852-0cac31e2-2ced-44ca-8825-f2cab3676c67.gif) ![CleanShot 2022-04-03 at 08 26 39](https://user-images.githubusercontent.com/1824611/161414852-0cac31e2-2ced-44ca-8825-f2cab3676c67.gif) [ ](https://user-images.githubusercontent.com/1824611/161414852-0cac31e2-2ced-44ca-8825-f2cab3676c67.gif)
> 
> I ended up using custom data-modal-hide and data-modal-show to make this work. Too bad though the toggle doesn't work :/

Can you share the code you used to get around this issue please?

**Comment by zoltanszogyenyi** [comment]
Hey everyone,

I'll get on this for the next update in the next two weeks.

I agree with the general idea so we'll provide a fix for this.

Thanks for your patience!

PS: if anyone wants to contribute they are more than welcome. The code can be found inside the `modal.js` file.

**Comment by HPODEV** [comment]
new demo with latest css and js version: https://jsfiddle.net/vincenzo_hpo/z9scjduf/ , bug persist with v1.5.1

**Comment by Tyjow** [comment]
Hi guys,

a quick fix for me, so hope it help you too

```
onHide: () => {
    console.log('modal is hidden');
    document.querySelector('[modal-backdrop]').remove();
},
```

**Comment by marciotoshio** [comment]
Hi everyone, I'm facing this issue and I'm proposing a fix here https://github.com/themesberg/flowbite/pull/274.

**Comment by yaqshank92** [comment]
Still facing this issue. Anyone with a fix without involving JQuery?

**Comment by KinG-InFeT** [comment]
> Hi guys,
> 
> a quick fix for me, so hope it help you too
> 
> ```
> onHide: () => {
>     console.log('modal is hidden');
>     document.querySelector('[modal-backdrop]').remove();
> },
> ```

@yaqshank92 see this comment, use onHide event without jquery

**Comment by KinG-InFeT** [comment]
> Hey everyone,
> 
> I know that this has been long overdue, but with the `v1.6.0` update and introducing TypeScript I have finally addressed this issue by changing the `data-modal-toggle` to `data-modal-target` which will be initialised only once, and the you can use `data-modal-toggle/show/hide={modalId}` to search for the modal instance and toggle, show or hide it.
> 
> Here's a breakdown of the code that does this:
> 
> ```
> const getModalInstance = (id: string, instances: ModalInstance[]) => {
>     if (instances.some((modalInstance) => modalInstance.id === id)) {
>         return instances.find((modalInstance) => modalInstance.id === id);
>     }
>     return null;
> };
> 
> export function initModals() {
>     const modalInstances = [] as ModalInstance[];
> 
>     document.querySelectorAll('[data-modal-target]').forEach(($triggerEl) => {
>         const modalId = $triggerEl.getAttribute('data-modal-toggle');
>         const $modalEl = document.getElementById(modalId);
>         const placement = $modalEl.getAttribute('data-modal-placement');
>         const backdrop = $modalEl.getAttribute('data-modal-backdrop');
> 
>         if (!getModalInstance(modalId, modalInstances)) {
>             modalInstances.push({
>                 id: modalId,
>                 object: new Modal(
>                     $modalEl as HTMLElement,
>                     {
>                         placement: placement ? placement : Default.placement,
>                         backdrop: backdrop ? backdrop : Default.backdrop,
>                     } as ModalOptions
>                 ),
>             });
>         }
>     });
> 
>     document.querySelectorAll('[data-modal-toggle]').forEach(($triggerEl) => {
>         const $targetEl = document.getElementById(
>             $triggerEl.getAttribute('data-modal-toggle')
>         );
>         const modalId = $targetEl.id;
>         const modal: ModalInstance = getModalInstance(modalId, modalInstances);
> 
>         if (modal) {
>             $triggerEl.addEventListener('click', () => {
>                 modal.object.toggle();
>             });
>         }
>     });
> 
>     document.querySelectorAll('[data-modal-show]').forEach(($triggerEl) => {
>         const $targetEl = document.getElementById(
>             $triggerEl.getAttribute('data-modal-show')
>         );
>         const modalId = $targetEl.id;
>         const modal: ModalInstance = getModalInstance(modalId, modalInstances);
> 
>         if (modal) {
>             $triggerEl.addEventListener('click', () => {
>                 if (modal.object.isHidden) {
>                     modal.object.show();
>                 }
>             });
>         }
>     });
> 
>     document.querySelectorAll('[data-modal-hide]').forEach(($triggerEl) => {
>         const $targetEl = document.getElementById(
>             $triggerEl.getAttribute('data-modal-hide')
>         );
>         const modalId = $targetEl.id;
>         const modal: ModalInstance = getModalInstance(modalId, modalInstances);
> 
>         if (modal) {
>             $triggerEl.addEventListener('click', () => {
>                 if (modal.object.isVisible) {
>                     modal.object.hide();
>                 }
>             });
>         }
>     });
> }
> ```
> 
> The update will arrive shortly and unfortunately we will have some breaking changes because `data-modal-toggle` will no longer initialise a modal object, only `data-modal-target`. The code will not break though.
> 
> This also fixed stacking up the event listeners for when click outside of the modal + escape button.
> 
> The functionality can be tested on this branch: https://github.com/themesberg/flowbite/tree/1.6.0

@zoltanszogyenyi  BUG persist : https://jsfiddle.net/vincenzo_hpo/z9scjduf/

**Comment by zoltanszogyenyi** [comment]
Hey @KinG-InFeT,

I've checked out the JS Fiddle example - but it looks like the modal works as intended.

Am I missing something here?

Thanks!

**Comment by KinG-InFeT** [comment]
> Hey @KinG-InFeT,
> 
> I've checked out the JS Fiddle example - but it looks like the modal works as intended.
> 
> Am I missing something here?
> 
> Thanks!

it opens, but if you close it normally, the black background remains.

**Comment by HPODEV** [comment]
bug persist with 1.6.2

**Comment by zoltanszogyenyi** [comment]
> > Hey @KinG-InFeT,
> > I've checked out the JS Fiddle example - but it looks like the modal works as intended.
> > Am I missing something here?
> > Thanks!
> 
> it opens, but if you close it normally, the black background remains.

The problem with your example is that you try to use both the data attributes method and create another instance for the same modal - we might implement a feature in the future where you can get the instance but for now that is not possible.

Please use either the data attributes or the programatic method, meaning if you want to create the class yourself you'll have to attach the event listeners yourself to the close button or toggle button.

**Comment by zoltanszogyenyi** [comment]
I will re-open this issue but I will label it as an enhancement for now because we haven't thought about allowing access to the instances that are being created via the data attributes.

**Comment by lincond** [comment]
Any updates to allowing acces to the instances that are being created via the data attributes? Thanks!

**Comment by Galanthus** [comment]
Hi all,

I am sharing my code I have it working when you click outside the Modal/Drawer to close and remove the backdrop classes. Feel free to test it.

```
import { Drawer } from 'flowbite';
import type { DrawerOptions, DrawerInterface } from 'flowbite';

const Menu = () => {
   const $targetEl = document.querySelector<HTMLElement>('#pvs-drawer-menu')!;

   const showButton = document.querySelector<HTMLElement>(
      '#drawer-show-button'
   )!;

   const closeButton = document.querySelector<HTMLElement>(
      '#drawer-hide-button'
   )!;

   const removeAllBackdrops = () => {
      const backdrops =
         document.querySelectorAll<HTMLElement>('[drawer-backdrop]');
      backdrops.forEach((backdrop) => backdrop.remove());
   };

   const options: DrawerOptions = {
      placement: 'right',
      backdrop: true,
      bodyScrolling: false,
      edge: false,
      edgeOffset: '',
      backdropClasses:
         'bg-[#131313] backdrop-blur-[2px] bg-opacity-70 fixed inset-0 z-30',
      onHide: () => {
         removeAllBackdrops();
      }
   };

   const drawer: DrawerInterface = new Drawer($targetEl, options);

   document.addEventListener('mouseup', (e: MouseEvent) => {
      const { target } = e;

      if (drawer.isVisible() && target != $targetEl) {
         drawer.hide();
         removeAllBackdrops();
      }
   });

   showButton.addEventListener('click', () => drawer.show());

   closeButton.addEventListener('click', () => drawer.hide());
};

document.addEventListener('DOMContentLoaded', Menu);
```

**Comment by sandeep728** [comment]
Any updates to allowing access to the instances that are being created via the data attributes? Thanks!

**Comment by alfonsrv** [comment]
Why not just add a `document.querySelector('#foo').modal` `property ` to the initialized modal element?

**Comment by SakNoelCode** [comment]
> > // show the modal
> > modal.show();
> 
> creating a modal object with options and using modal.show() and modal.hide() instead of the default data-modal-toggle, worked for me, tnks :D

It works for me too, here is my sample code:

Button code that is initialized via **data-modal-target**

`<button id="openModal" data-modal-target="popup-modal" class="block text-white bg-blue-700 hover:bg-blue-800 focus:ring-4 focus:outline-none focus:ring-blue-300 font-medium rounded-lg text-sm px-5 py-2.5 text-center dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800" type="button">
            Toggle modal
        </button>

        <div id="popup-modal" tabindex="-1" class="fixed top-0 left-0 right-0 z-50 hidden p-4 overflow-x-hidden overflow-y-auto md:inset-0 h-[calc(100%-1rem)] max-h-full">
            <div class="relative w-full max-w-md max-h-full">
                <div class="relative bg-white rounded-lg shadow dark:bg-gray-700">
                    <button type="button" class="absolute top-3 right-2.5 text-gray-400 bg-transparent hover:bg-gray-200 hover:text-gray-900 rounded-lg text-sm w-8 h-8 ml-auto inline-flex justify-center items-center dark:hover:bg-gray-600 dark:hover:text-white" onclick="closeModal()">
                        <svg class="w-3 h-3" aria-hidden="true" xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 14 14">
                            <path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="m1 1 6 6m0 0 6 6M7 7l6-6M7 7l-6 6" />
                        </svg>
                        <span class="sr-only">Close modal</span>
                    </button>
                    <div class="p-6 text-center">
                        <svg class="mx-auto mb-4 text-gray-400 w-12 h-12 dark:text-gray-200" aria-hidden="true" xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 20 20">
                            <path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M10 11V6m0 8h.01M19 10a9 9 0 1 1-18 0 9 9 0 0 1 18 0Z" />
                        </svg>
                        <h3 class="mb-5 text-lg font-normal text-gray-500 dark:text-gray-400">Are you sure you want to delete this product?</h3>
                        <button id="closeModal" type="button" class="text-white bg-red-600 hover:bg-red-800 focus:ring-4 focus:outline-none focus:ring-red-300 dark:focus:ring-red-800 font-medium rounded-lg text-sm inline-flex items-center px-5 py-2.5 text-center mr-2">
                            Yes, I'm sure
                        </button>
                        <button onclick="closeModal()" type="button" class="text-gray-500 bg-white hover:bg-gray-100 focus:ring-4 focus:outline-none focus:ring-gray-200 rounded-lg border border-gray-200 text-sm font-medium px-5 py-2.5 hover:text-gray-900 focus:z-10 dark:bg-gray-700 dark:text-gray-300 dark:border-gray-500 dark:hover:text-white dark:hover:bg-gray-600 dark:focus:ring-gray-600">No, cancel</button>
                    </div>
                </div>
            </div>
        </div>`

JavaScript Code:
`<script>
    const targetEl = document.getElementById('popup-modal');

    // options with default values
    const options = {
        placement: 'bottom-right',
        backdrop: 'dynamic',
        backdropClasses: 'bg-gray-900 bg-opacity-50 dark:bg-opacity-80 fixed inset-0 z-40',
        closable: true,
        onHide: () => {
            console.log('modal is hidden');
        },
        onShow: () => {
            console.log('modal is shown');
        },
        onToggle: () => {
            console.log('modal has been toggled');
        }
    };

    /*
     * targetEl: required
     * options: optional
     */
    const modal = new Modal(targetEl);

    document.getElementById('closeModal').addEventListener('click', function() {
        modal.hide();
    })

    document.getElementById('openModal').addEventListener('click', function() {
        modal.show();
    })

    function closeModal() {
        modal.hide();
    }

    function openModal() {
        modal.show();
    }
</script>`

**Comment by SakNoelCode** [comment]
Hi, I share the code that worked for me:

Button code:
`<button id="openModal" data-modal-target="popup-modal" type="button">
            Toggle modal
        </button>`

Modal Code:
`<div id="popup-modal" tabindex="-1" class="fixed top-0 left-0 right-0 z-50 hidden p-4 overflow-x-hidden overflow-y-auto md:inset-0 h-[calc(100%-1rem)] max-h-full">
            <div class="relative w-full max-w-md max-h-full">
                <div class="relative bg-white rounded-lg shadow dark:bg-gray-700">
                    <button type="button" class="absolute top-3 right-2.5 text-gray-400 bg-transparent hover:bg-gray-200 hover:text-gray-900 rounded-lg text-sm w-8 h-8 ml-auto inline-flex justify-center items-center dark:hover:bg-gray-600 dark:hover:text-white" onclick="closeModal()">
                        <svg class="w-3 h-3" aria-hidden="true" xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 14 14">
                            <path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="m1 1 6 6m0 0 6 6M7 7l6-6M7 7l-6 6" />
                        </svg>
                        <span class="sr-only">Close modal</span>
                    </button>
                    <div class="p-6 text-center">
                        <svg class="mx-auto mb-4 text-gray-400 w-12 h-12 dark:text-gray-200" aria-hidden="true" xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 20 20">
                            <path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M10 11V6m0 8h.01M19 10a9 9 0 1 1-18 0 9 9 0 0 1 18 0Z" />
                        </svg>
                        <h3 class="mb-5 text-lg font-normal text-gray-500 dark:text-gray-400">Are you sure you want to delete this product?</h3>
                        <button id="closeModal" type="button" class="text-white bg-red-600 hover:bg-red-800 focus:ring-4 focus:outline-none focus:ring-red-300 dark:focus:ring-red-800 font-medium rounded-lg text-sm inline-flex items-center px-5 py-2.5 text-center mr-2">
                            Yes, I'm sure
                        </button>
                        <button onclick="closeModal()" type="button" class="text-gray-500 bg-white hover:bg-gray-100 focus:ring-4 focus:outline-none focus:ring-gray-200 rounded-lg border border-gray-200 text-sm font-medium px-5 py-2.5 hover:text-gray-900 focus:z-10 dark:bg-gray-700 dark:text-gray-300 dark:border-gray-500 dark:hover:text-white dark:hover:bg-gray-600 dark:focus:ring-gray-600">No, cancel</button>
                    </div>
                </div>
            </div>
        </div>`

JS Code:
`const target = document.getElementById('popup-modal');
    
    const modal = new Modal(target);

    document.getElementById('closeModal').addEventListener('click', function() {
        modal.hide();
    })

    document.getElementById('openModal').addEventListener('click', function() {
        modal.show();
    })

    function closeModal() {
        modal.hide();
    }

    function openModal() {
        modal.show();
    }`

**Comment by KinG-InFeT** [comment]
in any case, the solution is only palliative, it is truly absurd that after months we still haven't arrived at a concrete and stable solution to the problem.

**Comment by alfonsrv** [comment]
![asd](https://github.com/themesberg/flowbite/assets/48770755/86704a1f-4cc2-48bc-ab9b-96349b22ab4d)

**Comment by rmelchner** [comment]
So in my case i ran into the Problem because i can't set the custom backdropClasses in data-attributes. 

That means im forced to initialise the Modal with JS and when i show it with modal.show() and close it the backdrop is not removed.

So from what i understand is that:

Either the onHide function needs to always trigger removal of the backdrop when you click a Close Button or backdropClasses should be available as a data-attribute.

Am i understanding the Problem correctly?

I created a PR for that:

https://github.com/themesberg/flowbite/pull/649

**Comment by KinG-InFeT** [comment]
> So in my case i ran into the Problem because i can't set the custom backdropClasses in data-attributes.
> 
> That means im forced to initialise the Modal with JS and when i show it with modal.show() and close if i close it by Clicking on a Close button inside the Modal - the backdrop is not removed.
> 
> So from what i understand is that:
> 
> Either the function (event Listener on data-modal-hide) needs to always trigger removal of the backdrop when you click a Close Button or backdropClasses should be available as a data-attribute or in the tailwind config file as default values.
> 
> Am i understanding the Problem correctly?
> 
> I created a PR for the first Problem (data-modal-hide):
> 
> #649
> 
> For the backdropClasses i created an issue:
> 
> #650

unfortunately this solution is not complete, if you open multiple modals, you need to have a link/id between the background and the modal so as to identify the modal that you may need to close.

**Comment by rmelchner** [comment]
@KinG-InFeT yes that would be great but at the moment it seems that modal-backdrop has no relation to the current modal anyways. Seems like the intended use case is to show only one modal at a time?

In theory it is possible i just tested it. Unfortunetly i don't have time to develop such a feature now. But it should not be hard since the Backdrop is added when a Modal is initialised (or open?).

One could add a data attribute with the id of the modal then in the data-modal-hide function we could find it the same way we find the modal.

**Comment by zoltanszogyenyi** [comment]
Hey everyone,

Just want to let you all know that I will provide a solution for this in the upcoming major version together with introducing prefixes - please stand by on this one. Appreciate the patience and if any one of you want to make a PR for this I'll be more than happy to have a look at it and even collaborate on it.

I'm currently working on the autocomplete component and right after that I'll prepare for these two new changes.

Cheers,
Zoltan

**Comment by rmelchner** [comment]
@zoltanszogyenyi thank you - i updated my PR https://github.com/themesberg/flowbite/pull/649

**Comment by rlongre** [comment]
@rmelchner This is a good solution for this particular issue. I added one comment but otherwise, I hope your PR is accepted.

I think there are still some underlying issues resulting from side effects of the init functions (initModals, initDrawers, etc.) being called automatically regardless of whether initFlowbite() is called. For example:
1.  The init functions should be made to be idempotent.
2. The public methods of components (show, hide, etc.) and the private methods they call (_createBackdrop, _destroyBackdropEl, etc.) should be stateful and not duplicate things that have already been done
3. If instances are going to be tracked at all, its scope shouldn't be limited to a closure within the init functions. This makes it impossible to determine whether that instance is being already tracked by a previous call.
4. Access to the object instances created via data attributes should be given. My third point would probably be necessary to facilitate that.

**Comment by s-aska** [comment]
I circumvented it as follows.

```ts
const close = element.querySelector('button[data-modal-hide]') as HTMLButtonElement
if (close) {
  close.click()
}
```

**Comment by DanielOberlechner** [comment]
This looks in my opinion as a very simple bug and I wonder why it consists after 1 1/2 years. This is my first time using this library but it doesn't look good to me in my opinion. This makes the whole component unusable without a hackfix.

Next time I think I'll probably stick with bootstrap or tailwindcss ui elements or even the html dialog element. I thought your library takes away the pain of styling elements but in the end its a mess.

And this library has also a commercial option which people probably rely on ... 

Gotta say I'm not really amazed at the moment, but no bad blood to the dev's ...

**Comment by zoltanszogyenyi** [comment]
Hey everyone!

Flowbite v2.0 fixes this issue by bringing a brand new Instances Manager.

Please read more about it here:

https://flowbite.com/docs/getting-started/javascript/index.html
https://flowbite.com/docs/getting-started/javascript/index.html

Cheers,
Zoltan

**Comment by KinG-InFeT** [comment]
> Hey everyone!
> 
> Flowbite v2.0 fixes this issue by bringing a brand new Instances Manager.
> 
> Please read more about it here:
> 
> https://flowbite.com/docs/getting-started/javascript/ https://github.com/themesberg/flowbite/releases/tag/v2.0.0
> 
> Cheers, Zoltan

finaly!, but out new problem, the modal not centered correctlly, here example: https://jsfiddle.net/vincenzo_hpo/r3ncmk6f/

**Comment by rmelchner** [comment]
> > Hey everyone!
> > Flowbite v2.0 fixes this issue by bringing a brand new Instances Manager.
> > Please read more about it here:
> > https://flowbite.com/docs/getting-started/javascript/ https://github.com/themesberg/flowbite/releases/tag/v2.0.0
> > Cheers, Zoltan
> 
> finaly!, but out new problem, the modal not centered correctlly, here example: https://jsfiddle.net/vincenzo_hpo/r3ncmk6f/

@KinG-InFeT change to "placement":"center" in the options

**Comment by KinG-InFeT** [comment]
> > > Hey everyone!
> > > Flowbite v2.0 fixes this issue by bringing a brand new Instances Manager.
> > > Please read more about it here:
> > > https://flowbite.com/docs/getting-started/javascript/ https://github.com/themesberg/flowbite/releases/tag/v2.0.0
> > > Cheers, Zoltan
> > 
> > 
> > finaly!, but out new problem, the modal not centered correctlly, here example: https://jsfiddle.net/vincenzo_hpo/r3ncmk6f/
> 
> @KinG-InFeT change to "placement":"center" in the options

oh sorry, yes it's work. thank you
